File: src/hardlaw/statute.py
"""Statute definitions — the "law on the books."

A Statute is a named hard rule that defines:
  - What constitutes a violation
  - What evidence is required
  - What happens on repeated violations
  - Default judgment posture (default-to-reject)
"""

from dataclasses import dataclass, field
from typing import Any, Literal


@dataclass
class ViolationType:
    """A kind of violation that can be found by an LLM judge.

    Attributes:
        name: Short identifier, e.g. "hate_speech", "missing_consent".
        severity: How serious the violation is.
        description: Human-readable definition for the judge.
    """

    name: str
    severity: Literal["critical", "high", "medium", "low"]
    description: str = ""


@dataclass
class EscalationRule:
    """What happens on repeated violations.

    Attributes:
        max_violations: Escalate after N violations of the same type.
        action: What action to take on escalation.
        escalate_to: Name of escalated statute, if escalation changes the rule set.
    """

    max_violations: int = 3
    action: Literal["block", "flag", "notify", "pause"] = "block"
    escalate_to: str | None = None


# ── Evidence Requirement (举证责任分层, mirrors Swift EvidenceRequirement) ──

from enum import Enum as _Enum

class EvidenceHolder(_Enum):
    WORKER = "worker"
    EMPLOYER = "employer"
    THIRD_PARTY = "third_party"

class MissingEvidenceAction(_Enum):
    FLAG = "flag"
    PROMPT = "prompt"
    BLOCK = "block"

@dataclass
class EvidenceRequirement:
    """A structured evidence requirement (v2, mirrors Swift)."""
    evidence: str
    holder: EvidenceHolder = EvidenceHolder.WORKER
    on_missing: MissingEvidenceAction = MissingEvidenceAction.BLOCK
    burden_basis: str | None = None
    alternatives: list["EvidenceRequirement"] = field(default_factory=list)
    min_count: int = 1

    def __post_init__(self):
        # Allow flexible construction from string (backward compat)
        pass

    @staticmethod
    def from_string(s: str) -> "EvidenceRequirement":
        return EvidenceRequirement(evidence=s)

    @staticmethod
    def from_list(items: list) -> list["EvidenceRequirement"]:
        """Parse mixed list of strings/dicts/EvidenceRequirement."""
        result = []
        for item in items:
            if isinstance(item, EvidenceRequirement):
                result.append(item)
            elif isinstance(item, str):
                result.append(EvidenceRequirement.from_string(item))
            elif isinstance(item, dict):
                result.append(EvidenceRequirement(
                    evidence=item.get("evidence", ""),
                    holder=EvidenceHolder(item.get("holder", "worker")),
                    on_missing=MissingEvidenceAction(item.get("on_missing", "block")),
                    burden_basis=item.get("burden_basis"),
                    alternatives=[EvidenceRequirement.from_string(a) for a in item.get("alternatives", [])],
                    min_count=item.get("min_count", 1),
                ))
        return result


@dataclass
class Statute:
    """A named hard rule encoding enforceable constraints on LLM agents.

    A Statute is data, not code — it can be serialized to JSON/YAML so that
    non-programmers can define the rules.

    Attributes:
        name: Unique identifier.
        description: Human-readable explanation.
        threshold: Multi-dimensional thresholds.
        required_evidence: Evidence requirements (v2: structured; v1: plain strings).
        violations: Kinds of violations this statute covers.
        escalation: What happens on repeated violations.
        default_to_reject: When the judge is uncertain, default to rejecting.
        blocking: If True, a violation blocks the action entirely.
    """

    name: str
    description: str = ""
    threshold: dict[str, Any] = field(default_factory=dict)
    required_evidence: list[EvidenceRequirement] = field(default_factory=list)
    violations: list[ViolationType] = field(default_factory=list)
    escalation: EscalationRule = field(default_factory=EscalationRule)
    default_to_reject: bool = True
    blocking: bool = True

    def to_dict(self) -> dict[str, Any]:
        """Serialize to a plain dict (JSON-compatible)."""
        return {
            "name": self.name,
            "description": self.description,
            "threshold": self.threshold,
            "required_evidence": self.required_evidence,
            "violations": [
                {"name": v.name, "severity": v.severity, "description": v.description}
                for v in self.violations
            ],
            "escalation": {
                "max_violations": self.escalation.max_violations,
                "action": self.escalation.action,
                "escalate_to": self.escalation.escalate_to,
            },
            "default_to_reject": self.default_to_reject,
            "blocking": self.blocking,
        }

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> "Statute":
        """Deserialize from a plain dict."""
        violations = [
            ViolationType(
                name=v["name"],
                severity=v["severity"],
                description=v.get("description", ""),
            )
            for v in data.get("violations", [])
        ]
        esc_data = data.get("escalation", {})
        escalation = EscalationRule(
            max_violations=esc_data.get("max_violations", 3),
            action=esc_data.get("action", "block"),
            escalate_to=esc_data.get("escalate_to"),
        )
        return cls(
            name=data["name"],
            description=data.get("description", ""),
            threshold=data.get("threshold", {}),
            required_evidence=EvidenceRequirement.from_list(data.get("required_evidence", [])),
            violations=violations,
            escalation=escalation,
            default_to_reject=data.get("default_to_reject", True),
            blocking=data.get("blocking", True),
        )


class StatuteBook:
    """A collection of statutes — the "legal code" for a domain.

    Provides lookup, validation, and serialization for multiple statutes.
    """

    def __init__(self, statutes: list[Statute] | None = None):
        self._statutes: dict[str, Statute] = {}
        if statutes:
            for s in statutes:
                self.add(s)

    def add(self, statute: Statute) -> None:
        """Add a statute to the book. Replaces if name already exists."""
        self._statutes[statute.name] = statute

    def get(self, name: str) -> Statute | None:
        """Look up a statute by name."""
        return self._statutes.get(name)

    def get_all(self, names: list[str]) -> list[Statute]:
        """Look up multiple statutes. Silently skips unknown names."""
        return [s for s in (self._statutes.get(n) for n in names) if s is not None]

    def list_names(self) -> list[str]:
        """Return all statute names."""
        return list(self._statutes.keys())

    def __len__(self) -> int:
        return len(self._statutes)

    def __contains__(self, name: str) -> bool:
        return name in self._statutes

    def to_dict(self) -> dict[str, Any]:
        """Serialize all statutes."""
        return {"statutes": [s.to_dict() for s in self._statutes.values()]}

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> "StatuteBook":
        """Deserialize a statute book."""
        statutes = [Statute.from_dict(s) for s in data.get("statutes", [])]
        return cls(statutes)

File: ios/HardlawKit/Sources/Statute/Statute.swift
import Foundation

// MARK: - ViolationType

/// A kind of violation that can be found by an LLM judge.
/// Mirrors Python `hardlaw.statute.ViolationType`.
public struct ViolationType: Codable, Equatable, Sendable {
    /// Short identifier, e.g. "hate_speech", "missing_consent".
    public var name: String
    /// How serious the violation is.
    public var severity: Severity
    /// Human-readable definition for the judge.
    public var description: String

    public init(name: String, severity: Severity, description: String = "") {
        self.name = name
        self.severity = severity
        self.description = description
    }
}

// MARK: - Severity

/// Violation severity level.
/// Mirrors Python `Literal["critical", "high", "medium", "low"]`.
public enum Severity: String, Codable, Sendable, CaseIterable {
    case critical, high, medium, low
}

// MARK: - EscalationRule

/// What happens on repeated violations.
/// Mirrors Python `hardlaw.statute.EscalationRule`.
public struct EscalationRule: Codable, Equatable, Sendable {
    /// Escalate after N violations of the same type.
    public var maxViolations: Int
    /// What action to take on escalation.
    public var action: EscalationAction
    /// Name of escalated statute, if escalation changes the rule set.
    public var escalateTo: String?

    public init(maxViolations: Int = 3, action: EscalationAction = .block, escalateTo: String? = nil) {
        self.maxViolations = maxViolations
        self.action = action
        self.escalateTo = escalateTo
    }

    // Map Python JSON keys: "max_violations", "action", "escalate_to"
    enum CodingKeys: String, CodingKey {
        case maxViolations = "max_violations"
        case action
        case escalateTo = "escalate_to"
    }
}

/// Escalation actions matching Python `Literal["block", "flag", "notify", "pause"]`.
public enum EscalationAction: String, Codable, Sendable, CaseIterable {
    case block, flag, notify, pause
}

// MARK: - Evidence Requirement (举证责任分层)

/// Who holds the evidence — determines what happens when it's missing.
public enum EvidenceHolder: String, Codable, Sendable, CaseIterable {
    case worker     // 劳动者应自行举证
    case employer   // 用人单位掌握管理（《劳动争议调解仲裁法》第6条）
    case thirdParty // 行政机关/第三方，可申请调取
}

/// What to do when required evidence is missing.
public enum MissingEvidenceAction: String, Codable, Sendable, CaseIterable {
    case flag    // 提示缺口+举证责任提示，不阻断
    case prompt  // 要求补充或申请调取后继续
    case block   // 阻断，无重试（劳动者自行举证/LLM幻觉）
}

/// A structured evidence requirement replacing flat `[String]` lists.
/// Each entry specifies who should provide the evidence, what happens
/// when it's missing, and what alternatives exist.
public struct EvidenceRequirement: Codable, Equatable, Sendable {
    /// Evidence name, e.g. "工资表"
    public var evidence: String
    /// Who should provide this evidence
    public var holder: EvidenceHolder
    /// Action when evidence is missing
    public var onMissing: MissingEvidenceAction
    /// Legal basis for burden assignment
    public var burdenBasis: String?
    /// Alternative evidence that satisfies the same requirement
    public var alternatives: [EvidenceRequirement]
    /// Minimum number of alternatives (including primary) that must be satisfied
    public var minCount: Int

    public init(
        evidence: String,
        holder: EvidenceHolder = .worker,
        onMissing: MissingEvidenceAction = .block,
        burdenBasis: String? = nil,
        alternatives: [EvidenceRequirement] = [],
        minCount: Int = 1
    ) {
        self.evidence = evidence
        self.holder = holder
        self.onMissing = onMissing
        self.burdenBasis = burdenBasis
        self.alternatives = alternatives
        self.minCount = minCount
    }

    /// Convenience: create a simple worker-held requirement (backward compat).
    public init(_ evidence: String) {
        self.evidence = evidence
        self.holder = .worker
        self.onMissing = .block
        self.burdenBasis = nil
        self.alternatives = []
        self.minCount = 1
    }
}

// MARK: - Statute

/// A named hard rule encoding enforceable constraints on LLM agents.
/// A Statute is data, not code — it can be serialized to JSON/YAML.
/// Mirrors Python `hardlaw.statute.Statute`.
public struct Statute: Codable, Equatable, Sendable {
    /// Unique identifier, e.g. "hate_speech", "gdpr_consent".
    public var name: String
    /// Human-readable explanation of what this statute governs.
    public var description: String
    /// Multi-dimensional thresholds, e.g. {"confidence_min": "medium"}.
    public var threshold: [String: JSONValue]
    /// Evidence requirements (v2: structured with burden layering).
    /// Backward compatible: old `[String]` decodes to `.worker` + `.block`.
    public var requiredEvidence: [EvidenceRequirement]
    /// Kinds of violations this statute covers.
    public var violations: [ViolationType]
    /// What happens on repeated violations.
    public var escalation: EscalationRule
    /// When the judge is uncertain, default to rejecting.
    public var defaultToReject: Bool
    /// If True, a violation blocks the action entirely (no retry).
    public var blocking: Bool

    public init(
        name: String,
        description: String = "",
        threshold: [String: JSONValue] = [:],
        requiredEvidence: [EvidenceRequirement] = [],
        violations: [ViolationType] = [],
        escalation: EscalationRule = EscalationRule(),
        defaultToReject: Bool = true,
        blocking: Bool = true
    ) {
        self.name = name
        self.description = description
        self.threshold = threshold
        self.requiredEvidence = requiredEvidence
        self.violations = violations
        self.escalation = escalation
        self.defaultToReject = defaultToReject
        self.blocking = blocking
    }

    // Map Python JSON keys exactly (snake_case)
    enum CodingKeys: String, CodingKey {
        case name, description, threshold
        case requiredEvidence = "required_evidence"
        case violations, escalation
        case defaultToReject = "default_to_reject"
        case blocking
    }

    // Custom decoding with backward compatibility for old `[String]` format
    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        name = try container.decode(String.self, forKey: .name)
        description = try container.decodeIfPresent(String.self, forKey: .description) ?? ""
        threshold = try container.decodeIfPresent([String: JSONValue].self, forKey: .threshold) ?? [:]
        // Backward compat: try new [EvidenceRequirement] format first,
        // fall back to old [String] format (each string → .worker + .block)
        if let reqs = try? container.decode([EvidenceRequirement].self, forKey: .requiredEvidence) {
            requiredEvidence = reqs
        } else if let strings = try? container.decode([String].self, forKey: .requiredEvidence) {
            requiredEvidence = strings.map { EvidenceRequirement($0) }
        } else {
            requiredEvidence = []
        }
        violations = try container.decodeIfPresent([ViolationType].self, forKey: .violations) ?? []
        escalation = try container.decodeIfPresent(EscalationRule.self, forKey: .escalation) ?? EscalationRule()
        defaultToReject = try container.decodeIfPresent(Bool.self, forKey: .defaultToReject) ?? true
        blocking = try container.decodeIfPresent(Bool.self, forKey: .blocking) ?? true
    }

    /// Serialize to a plain dict (JSON-compatible, mirrors Python `to_dict`).
    public func toDict() -> [String: Any] {
        var result: [String: Any] = [
            "name": name,
            "description": description,
            "threshold": threshold.mapValues { $0 },
            "required_evidence": requiredEvidence.map { req in
                var d: [String: Any] = ["evidence": req.evidence, "holder": req.holder.rawValue, "on_missing": req.onMissing.rawValue]
                if let b = req.burdenBasis { d["burden_basis"] = b }
                if !req.alternatives.isEmpty { d["alternatives"] = req.alternatives.map { $0.evidence } }
                return d
            },
            "violations": violations.map { v in
                [
                    "name": v.name,
                    "severity": v.severity.rawValue,
                    "description": v.description,
                ]
            },
            "escalation": [
                "max_violations": escalation.maxViolations,
                "action": escalation.action.rawValue,
                "escalate_to": escalation.escalateTo as Any,
            ],
            "default_to_reject": defaultToReject,
            "blocking": blocking,
        ]
        // remove NSNull for escalate_to when nil (matches Python behavior)
        if escalation.escalateTo == nil {
            if var esc = result["escalation"] as? [String: Any] {
                esc.removeValue(forKey: "escalate_to")
                result["escalation"] = esc
            }
        }
        return result
    }

    /// Deserialize from a plain dict (mirrors Python `from_dict`).
    public static func fromDict(_ data: [String: Any]) -> Statute {
        let violations: [ViolationType] = (data["violations"] as? [[String: Any]] ?? []).map { v in
            ViolationType(
                name: v["name"] as? String ?? "",
                severity: Severity(rawValue: v["severity"] as? String ?? "low") ?? .low,
                description: v["description"] as? String ?? ""
            )
        }
        let escData = data["escalation"] as? [String: Any] ?? [:]
        let escalation = EscalationRule(
            maxViolations: escData["max_violations"] as? Int ?? 3,
            action: EscalationAction(rawValue: escData["action"] as? String ?? "block") ?? .block,
            escalateTo: escData["escalate_to"] as? String
        )
        return Statute(
            name: data["name"] as? String ?? "",
            description: data["description"] as? String ?? "",
            threshold: (data["threshold"] as? [String: Any] ?? [:]).mapValues { parseJSONPrimitive($0) },
            requiredEvidence: parseRequiredEvidence(data["required_evidence"]),
            violations: violations,
            escalation: escalation,
            defaultToReject: data["default_to_reject"] as? Bool ?? true,
            blocking: data["blocking"] as? Bool ?? true
        )
    }
}

/// Recursively convert Any → JSONValue for fromDict.
private func parseJSONPrimitive(_ value: Any) -> JSONValue {
    switch value {
    case let s as String: return .string(s)
    case let n as Double: return .number(n)
    case let n as Int: return .number(Double(n))
    case let b as Bool: return .bool(b)
    case let arr as [Any]: return .array(arr.map(parseJSONPrimitive))
    case let dict as [String: Any]: return .object(dict.mapValues(parseJSONPrimitive))
    case is NSNull: return .null
    default: return .string("\(value)")
    }
}

/// Parse required_evidence from dict — handles both old `[String]` and new structured format.
private func parseRequiredEvidence(_ value: Any?) -> [EvidenceRequirement] {
    // New format: array of dicts
    if let dicts = value as? [[String: Any]] {
        return dicts.compactMap { d in
            guard let evidence = d["evidence"] as? String else { return nil }
            return EvidenceRequirement(
                evidence: evidence,
                holder: EvidenceHolder(rawValue: d["holder"] as? String ?? "worker") ?? .worker,
                onMissing: MissingEvidenceAction(rawValue: d["on_missing"] as? String ?? "block") ?? .block,
                burdenBasis: d["burden_basis"] as? String,
                alternatives: (d["alternatives"] as? [String])?.map { EvidenceRequirement($0) } ?? [],
                minCount: d["min_count"] as? Int ?? 1
            )
        }
    }
    // Old format: array of strings
    if let strings = value as? [String] {
        return strings.map { EvidenceRequirement($0) }
    }
    return []
}
